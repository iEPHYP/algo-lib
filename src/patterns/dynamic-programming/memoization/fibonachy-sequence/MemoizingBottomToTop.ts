// Complexity O(N), O(N)S
export const calcFibonachyItemMemoizingBottomToTopAt = (lastIndex: number) => {
  const cache: number[] = [1, 1];

  const getFibonachyItem = (index: number): number => {
    if (index < 2) {
      return 1;
    }

    return cache[index - 1] + cache[index - 2];
  };

  for (let i = 0; i <= lastIndex; i++) {
    cache[i] = getFibonachyItem(i);
  }

  return cache[lastIndex];
};
