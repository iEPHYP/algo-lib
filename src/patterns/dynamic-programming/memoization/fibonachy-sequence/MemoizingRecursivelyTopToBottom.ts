// Complexity O(N)T, O(2N)S
export const calcFibonachyItemMemoizingRecursivelyTopToBottomAt = (lastIndex: number) => {
  const cache: number[] = [1, 1];

  const getFibonachyItem = (index: number): number => {
    if (index < 2) {
      return 1;
    }
    if (cache[index]) {
      return cache[index];
    }

    // Not good, cuz can exceed call stack size, because too many function calls until the bottom
    cache[index] = getFibonachyItem(index - 1) + getFibonachyItem(index - 2);

    return cache[index];
  };

  return getFibonachyItem(lastIndex);
};
