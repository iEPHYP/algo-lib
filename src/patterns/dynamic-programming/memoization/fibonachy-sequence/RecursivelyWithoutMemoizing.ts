// Complexity O(2^n)
export const calcFibonachyItemRecursivelyAt = (index: number): number => {
  if (index <= 1) return 1;

  return calcFibonachyItemRecursivelyAt(index - 1) + calcFibonachyItemRecursivelyAt(index - 2);
};
