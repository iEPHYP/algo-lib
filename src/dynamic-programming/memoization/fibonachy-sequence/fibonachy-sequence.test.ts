import { calcFibonachyItemMemoizingBottomToTopAt } from './MemoizingBottomToTop';
import { calcFibonachyItemMemoizingRecursivelyTopToBottomAt } from './MemoizingRecursivelyTopToBottom';
import { calcFibonachyItemRecursivelyAt } from './RecursivelyWithoutMemoizing';

describe(calcFibonachyItemMemoizingRecursivelyTopToBottomAt.name, () => {
  it('works properly', () => {
    expect(calcFibonachyItemMemoizingRecursivelyTopToBottomAt(2)).toBe(2);
    expect(calcFibonachyItemMemoizingRecursivelyTopToBottomAt(3)).toBe(3);
    expect(calcFibonachyItemMemoizingRecursivelyTopToBottomAt(4)).toBe(5);
    expect(calcFibonachyItemMemoizingRecursivelyTopToBottomAt(5)).toBe(8);
    expect(calcFibonachyItemMemoizingRecursivelyTopToBottomAt(6)).toBe(13);
    expect(calcFibonachyItemMemoizingRecursivelyTopToBottomAt(30)).toMatchSnapshot();
    try {
      expect(calcFibonachyItemMemoizingRecursivelyTopToBottomAt(100000000)).toMatchSnapshot();
    } catch (e) {
      expect(e).toMatchSnapshot('Call stack exceeded error');
    }
  });
});

describe(calcFibonachyItemMemoizingBottomToTopAt.name, () => {
  it('works properly', () => {
    expect(calcFibonachyItemMemoizingBottomToTopAt(2)).toBe(2);
    expect(calcFibonachyItemMemoizingBottomToTopAt(3)).toBe(3);
    expect(calcFibonachyItemMemoizingBottomToTopAt(4)).toBe(5);
    expect(calcFibonachyItemMemoizingBottomToTopAt(5)).toBe(8);
    expect(calcFibonachyItemMemoizingBottomToTopAt(6)).toBe(13);
    expect(calcFibonachyItemMemoizingBottomToTopAt(30)).toMatchSnapshot();
    // expect(calcFibonachyItemBottomToTopAt(100000000)).toMatchSnapshot();
  });
});

describe(calcFibonachyItemRecursivelyAt.name, () => {
  it('works properly', () => {
    expect(calcFibonachyItemRecursivelyAt(2)).toBe(2);
    expect(calcFibonachyItemRecursivelyAt(3)).toBe(3);
    expect(calcFibonachyItemRecursivelyAt(4)).toBe(5);
    expect(calcFibonachyItemRecursivelyAt(5)).toBe(8);
    expect(calcFibonachyItemRecursivelyAt(6)).toBe(13);
    expect(calcFibonachyItemRecursivelyAt(30)).toMatchSnapshot();
  });
});
