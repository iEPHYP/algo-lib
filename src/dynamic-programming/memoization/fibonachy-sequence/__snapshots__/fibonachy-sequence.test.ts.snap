// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calcFibonachyItemMemoizingBottomToTopAt works properly 1`] = `1346269`;

exports[`calcFibonachyItemMemoizingRecursivelyTopToBottomAt works properly 1`] = `1346269`;

exports[`calcFibonachyItemMemoizingRecursivelyTopToBottomAt works properly: Call stack exceeded error 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`calcFibonachyItemRecursivelyAt works properly 1`] = `1346269`;
